pkg/util/v1beta1/katibconfig/config.go

```diff
@@ -42,6 +42,7 @@ type SuggestionConfig struct {
 	PersistentVolumeClaimSpec corev1.PersistentVolumeClaimSpec `json:"persistentVolumeClaimSpec,omitempty"`
 	PersistentVolumeSpec      corev1.PersistentVolumeSpec      `json:"persistentVolumeSpec,omitempty"`
 	PersistentVolumeLabels    map[string]string                `json:"persistentVolumeLabels,omitempty"`
+	VolumeForceMount          bool                             `json:"volumeForceMount,omitempty"`
```

Comment:
Can we avoid additional `VolumeForceMount` parameter in config ?

One suggestion that I have: Always [get  desired volume](https://github.com/kubeflow/katib/blob/3f1eba01d6f6b1520d81b7c71dbfb57c143b62d9/pkg/controller.v1beta1/suggestion/suggestion_controller.go#L202) if Suggestion Config Data contains Persistent Volume Spec or if Resume Policy = From Volume. 
The default values for PVC, we can set under [the Desired Volume](https://github.com/kubeflow/katib/blob/3f1eba01d6f6b1520d81b7c71dbfb57c143b62d9/pkg/controller.v1beta1/suggestion/composer/composer.go#L260) function.

I understand that it doesn't solve the problem when user wants to use the default values for Suggestion volume and still attach the volume, but we can find the better solution.

Any other ideas how to avoid more parameters in Katib Config @a9p @tenzen-y @johnugeorge ?

Reply:
> Always [get desired volume](https://github.com/kubeflow/katib/blob/3f1eba01d6f6b1520d81b7c71dbfb57c143b62d9/pkg/controller.v1beta1/suggestion/suggestion_controller.go#L202) if Suggestion Config Data contains Persistent Volume Spec or if Resume Policy = From Volume.
The default values for PVC, we can set under [the Desired Volume](https://github.com/kubeflow/katib/blob/3f1eba01d6f6b1520d81b7c71dbfb57c143b62d9/pkg/controller.v1beta1/suggestion/composer/composer.go#L260) function.

Maybe, it's idea is better to avoid more parameters in katib-config.

> Any other ideas how to avoid more parameters in Katib Config

I have no good ideas.

Reply:
I agree with removing the new parameter 👍🏽

I'm not sure what a good default is for the mounts or if there is one -- it seems like that would be setup dependent. One thought would be to prevent the suggestion service from starting if the mount doesn't exist in the config (making the config issue apparent up front) for services that require one? Currently, PBT for example, basically does a random search at the initial seed until run limits are hit which is a bit opaque unless you read the documentation for the config needed to operate correctly.

Reply:
>One thought would be to prevent the suggestion service from starting if the mount doesn't exist in the config (making the config issue apparent up front) for services that require one?

@a9p How we are going to understand for which Suggestion service the Volume is mandatory ?

Reply:
I was originally thinking something like [this](https://github.com/kubeflow/katib/blob/1e7e3c16fed97c3d3bb1c4e5fd48bae5ebd9401f/pkg/controller.v1beta1/suggestion/suggestion_controller.go#L50), but hardcoding this in the suggestion controller isn't ideal. Possibly linking the required resources to the suggestion service and reading that in at the controller (e.g. an api change to the rpc endpoint implemented by the suggestion defaulting to no additional requirements)?

Reply:
@a9p Or maybe we can use [`ValidateAlgorithmSettings` gRPC API](https://github.com/kubeflow/katib/blob/1e7e3c16fed97c3d3bb1c4e5fd48bae5ebd9401f/pkg/apis/manager/v1beta1/api.proto#L36) ?
In addition to the Experiment's Algorithm Setting we can pass information (with appropriate algorithm name) from the Katib config to the Suggestion service ?
Then, service can verify if settings are correct and user will see the results on the Experiment status.

What do you think @a9p @johnugeorge @tenzen-y @gaocegege ?

Reply:
Does something like the following match what you were thinking:
```
message AlgorithmSpec {
    string algorithm_name = 1;
    repeated AlgorithmSetting algorithm_settings = 2;
    repeated <ServiceSetting> <service>_settings = 3; // ServiceSetting, SuggestionSetting, etc. 
}
```

Reply:
@a9p @andreyvelich Looks good about expanding gRPC API.

> Any ideas how we should define SuggestionSetting API in our gRPC since we might want to send all [SuggestionConfig parameters to the Suggestion settings (which is Deployment spec)](https://github.com/kubeflow/katib/blob/master/pkg/util/v1beta1/katibconfig/config.go#L36)?

Maybe, to use all Deployment fields, we need to hold the [kubernetes/api](https://github.com/kubernetes/api) repository as git submodules and import deployment spec gRPC API to our gRPC API.

And then, we will need to manage to support K8s versions more carefully since the above means that our gRPC API intensely depends on K8s API.

@andreyvelich @a9p wdyt?

Reply:
Sounds great @tenzen-y , so should we [import the Deployment spec from the proto file](https://github.com/kubernetes/api/blob/master/apps/v1/generated.proto#L268) and make the following API structure:
```proto
message AlgorithmSpec {
    string algorithm_name = 1;
    repeated AlgorithmSetting algorithm_settings = 2;
    DeploymentSpec suggestion_spec = 3;
}
```
> And then, we will need to more carefully manage K8s versions supported by katib since the above means that our gRPC API intensely depends on K8s API.

I think we already use the Kubernetes API for our CRD API, so we already have to be compatible with the version (https://github.com/kubeflow/katib/blob/master/pkg/apis/controller/common/v1beta1/common_types.go#L233).
cc @johnugeorge

Reply:
sgtm

Reply:
Across most of the repo where the k8s spec is used, it is used in go via imports -- is git submoduling in the manager/v1beta1 directory the preferred approach? I put a temporary shim in the build.sh tying it to the minor k8s version just to verify static checks, but the code should still fail at runtime since it's deleting the directory.

Reply:
Hi @andreyvelich / @andreyvelich! Curious if either of you have any thoughts on the above?

Reply:
@a9p Apologies for the late reply, please can we discuss it again after the Kubeflow release (e.g. after 2 weeks) ?

Reply:
@andreyvelich / @tenzen-y -- just following up if either of you have any thoughts on the best way to include the k8s spec in this repo?

Reply:
Sorry for the late response. I'm missing the notifications. I came back to this thread right now.
I'll check your changes.

Reply:
> is git submoduling in the manager/v1beta1 directory the preferred approach? I put a temporary shim in the build.sh tying it to the minor k8s version just to verify static checks, but the code should still fail at runtime since it's deleting the directory.

@a9p Thanks for your investigating. 
Can we store the proto definition for the k8s.io/apps/v1 only when generating codes? I guess we can store the k8s proto file locally like the following.

https://github.com/kubeflow/mpi-operator/blob/c034b8caf613b2951aa2550c32f1333c58f56a39/Makefile#L170-L172

I guess that we can do that with `GOPATH=/tmp go install k8s.io/api`.
I appreciate your effort!